prov/tcp/src/tcpx_progress.c

```diff
@@ -846,3 +813,159 @@ static ssize_t (*tcpx_start_op[ofi_op_write + 1])(struct tcpx_ep *ep) = {
 	[ofi_op_read_rsp] = tcpx_op_read_rsp,
 	[ofi_op_write] = tcpx_op_write,
 };
+
+void tcpx_progress_domain(struct tcpx_domain *domain)
+{
+	struct ofi_epollfds_event events[MAX_POLL_EVENTS];
+	struct dlist_entry *item;
+	struct tcpx_ep *ep;
+	struct fid *fid;
+	uint32_t inevent, outevent, errevent;
+	int nfds, i;
+
+	ofi_mutex_lock(&domain->progress_lock);
+	dlist_foreach(&domain->ep_list, item) {
+		ep = container_of(item, struct tcpx_ep, ep_entry);
+
+		ofi_mutex_lock(&ep->lock);
+		/* We need to progress receives in the case where we're waiting
+		 * on the application to post a buffer to consume a receive
+		 * that we've already read from the kernel.  If the message is
+		 * of length 0, there's no additional data to read, so failing
+		 * to progress can result in application hangs.
+		 */
+		if (ofi_bsock_readable(&ep->bsock) ||
+		    (ep->cur_rx.handler && !ep->cur_rx.entry)) {
+			assert(ep->state == TCPX_CONNECTED);
+			tcpx_progress_rx(ep);
+		}
+
+		(void) tcpx_update_epoll(ep);
+		ofi_mutex_unlock(&ep->lock);
+	}
+	ofi_mutex_unlock(&domain->progress_lock);
```

Comment:
I found an interesting bug while testing the patch: the function keeps accessing an endpoint after it is closed.
The problem is that `domain->progress_lock` is released early in the function, then the function calls `ofi_{epoll/pollfds}_wait()` and processes ready events. If the endpoint is closed after the call to `ofi_{epoll/pollfds}_wait()` but before processing the events, the function may access a dead endpoint (e.g., the log `ep->lock` would already been destroyed).

One way to work around this problem is to keep the lock `domain->progress_lock` until the end of the function (like it was in the former code).

Reply:
Accessing the ep after its closed is an area I'm actively concerned about.  I'm not seeing an alternative to holding the progress_lock throughout the call.